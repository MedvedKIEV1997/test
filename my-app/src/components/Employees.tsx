import { useState, useEffect } from 'react';
import { useParams } from 'react-router-dom';
import { CopyToClipboard } from 'react-copy-to-clipboard';
import {
    Button,
    Toast,
    ToastContainer,
    Container,
    Placeholder
} from 'react-bootstrap';

import { MainText, SecondaryText } from '../styled/employees.styles';
import useTypedSelector from '../hooks/useTypedSelectors';
import {
    selectEmployee,
    selectLoading,
    selectSubordinatesToShow
} from '../redux/ducks/employee.duck';
import useActions from '../hooks/useActions';

const Employees = () => {
    const [show, setShow] = useState(false);
    const { id } = useParams();
    const url = window.location.href;
    const actions = useActions();
    const employee = useTypedSelector(selectEmployee);
    const subordinates = useTypedSelector(selectSubordinatesToShow);
    const isLoading = useTypedSelector(selectLoading);

    useEffect(() => {
        actions.getEmployee({ id: Number(id) });
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [id]);

    if (isLoading) {
        return (
            <Container className="my-5 position-relative">
                <Placeholder as={MainText} animation="glow">
                    <Placeholder xs={4} />
                </Placeholder>
                <Placeholder as={SecondaryText} animation="glow">
                    <Placeholder xs={3} />
                </Placeholder>
                <Placeholder as={MainText} animation="glow">
                    <Placeholder xs={4} />
                </Placeholder>
                <Placeholder as={SecondaryText} animation="glow">
                    <Placeholder xs={2} /> <Placeholder xs={3} />{' '}
                    <Placeholder xs={2} />
                </Placeholder>

                <Placeholder.Button
                    className="position-absolute top-0 end-0"
                    animation="glow"
                >
                    Copy this URL
                </Placeholder.Button>
            </Container>
        );
    } else {
        return (
            <Container className="my-5 position-relative">
                <MainText>Employee:</MainText>
                <SecondaryText>
                    {employee ? employee.name : 'Not Found'}{' '}
                </SecondaryText>

                {subordinates.length !== 0 && (
                    <>
                        <MainText>Subordinates:</MainText>
                        <SecondaryText>
                            {subordinates
                                .map((subordinate) => subordinate.name)
                                .join(', ')}
                        </SecondaryText>
                    </>
                )}
                {employee && (
                    <CopyToClipboard text={url} onCopy={() => setShow(true)}>
                        <Button className="position-absolute top-0 end-0">
                            Copy this URL
                        </Button>
                    </CopyToClipboard>
                )}
                <ToastContainer position={'bottom-start'} className="p-3">
                    <Toast
                        onClose={() => setShow(false)}
                        show={show}
                        delay={3000}
                        autohide
                    >
                        <Toast.Body>Url is copied to clipboard!</Toast.Body>
                    </Toast>
                </ToastContainer>
            </Container>
        );
    }
};

export default Employees;
