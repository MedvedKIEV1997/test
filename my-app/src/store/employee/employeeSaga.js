import { all, call, put, select, takeLatest } from 'redux-saga/effects';
import axios from 'axios';
import {
    getEmployeeFailure,
    getEmployeeFetch,
    getEmployeeSuccess,
    getSearchFailure,
    getSearchFetch,
    getSearchSuccess,
    getSubordinatesFailure,
    getSubordinatesFetch,
    getSubordinatesSuccess,
    selectSubordinates
} from './employeeSlice';

function* workGetSearchFetch({ payload }) {
    try {
        const response = yield call(
            axios.get,
            `http://localhost:8000/employees/search/${payload}`
        );
        yield put(getSearchSuccess(response.data));
    } catch (error) {
        yield put(getSearchFailure(error));
    }
}

function* workGetEmployeeFetch({ payload }) {
    try {
        const response = yield call(
            axios.get,
            `http://localhost:8000/employees/${payload}`
        );
        yield put(getEmployeeSuccess(response.data));
        yield call(workGetSubordinatesFetch);
    } catch (error) {
        yield put(getEmployeeFailure(error));
    }
}

function* workGetSubordinatesFetch() {
    try {
        const subordinates = yield select(selectSubordinates);
        if (subordinates.length !== 0) {
            const response = yield all(
                subordinates.map((id) => {
                    const response = call(
                        axios.get,
                        `http://localhost:8000/employees/${id}`
                    );
                    return response;
                })
            );

            const subordinatesData = response.map((el) => el.data);
            console.log(subordinatesData);
            yield put(getSubordinatesSuccess(subordinatesData));
            yield call(workGetSubordinatesFetch);
        }
    } catch (error) {
        yield put(getSubordinatesFailure(error));
    }
}

function* onSearchFetch() {
    yield takeLatest(getSearchFetch.type, workGetSearchFetch);
}

function* onEmployeeFetch() {
    yield takeLatest(getEmployeeFetch.type, workGetEmployeeFetch);
}

function* onSubordinatesFetch() {
    yield takeLatest(getSubordinatesFetch.type, workGetSubordinatesFetch);
}

export function* employeeSaga() {
    yield all([
        call(onEmployeeFetch),
        call(onSubordinatesFetch),
        call(onSearchFetch)
    ]);
}

export default employeeSaga;
