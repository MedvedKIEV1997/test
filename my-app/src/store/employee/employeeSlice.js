import { createSlice } from '@reduxjs/toolkit';

const employeeSlice = createSlice({
  name: 'employee',
  initialState: {
    employee: null,
    subordinates: [],
    subordinatesToShow: [],

    error: null,
  },
  reducers: {
    getEmployeeFetch: (state, action) => {
      state.employee = null;
      state.subordinates = [];
      state.error = null;
    },
    getEmployeeSuccess: (state, action) => {
      state.employee = action.payload;
      state.subordinates = action.payload.controls;
      state.subordinatesToShow = action.payload.controls;
    },
    getEmployeeFailure: (state, action) => {
      state.error = action.payload;
      state.subordinates = [];
      state.subordinatesToShow = [];
    },
    getSubordinatesFetch: (state, action) => {
      state.subordinatesToShow = state.subordinates;
    },
    getSubordinatesSuccess: (state, action) => {
      state.subordinatesToShow = [
        ...state.subordinatesToShow,
        ...action.payload,
      ];
      state.subordinates = action.payload;
    },
    getSubordinatesFailure: (state, action) => {
      state.error = action.payload;
      state.subordinates = [];
      state.subordinatesToShow = [];
    },
  },
});

export const {
  getEmployeeFailure,
  getEmployeeSuccess,
  getEmployeeFetch,
  getSubordinatesFailure,
  getSubordinatesFetch,
  getSubordinatesSuccess,
} = employeeSlice.actions;

export const selectEmployee = (state) => state.employee.employee;
export const selectSubordinates = (state) => state.employee.subordinates;
export const selectSubordinatesToShow = (state) =>
  state.employee.subordinatesToShow;

export default employeeSlice.reducer;
