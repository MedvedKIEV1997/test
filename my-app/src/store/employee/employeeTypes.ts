export type Employee = {
    name: string;
    controls: number[];
    id: number;
};

export type SearchEmployeesPayload = {
    name: string;
};

export type GetEmployeePayload = {
    id: number;
};
