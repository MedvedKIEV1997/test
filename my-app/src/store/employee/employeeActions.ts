import { createAction } from '@reduxjs/toolkit';
import { GetEmployeePayload, SearchEmployeesPayload } from './employeeTypes';

export const searchEmployees =
    createAction<SearchEmployeesPayload>('searchEmployees');

export const getEmployee = createAction<GetEmployeePayload>('getEmployee');

export const getSubordinates = createAction('getSubordinates');
