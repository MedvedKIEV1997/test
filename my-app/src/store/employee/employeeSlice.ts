import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../store';
import { Employee } from './employeeTypes';

type InitialState = {
    employee: Employee | null;
    subordinates: number[];
    subordinatesToShow: Employee[];
    search: Employee[];
};

const initialState: InitialState = {
    employee: null,
    subordinates: [],
    subordinatesToShow: [],
    search: []
};

const employeeSlice = createSlice({
    name: 'employee',
    initialState,
    reducers: {
        setSearch: (state, { payload }: PayloadAction<Employee[]>) => {
            state.search = payload;
        },

        setEmployee: (state, { payload }: PayloadAction<Employee>) => {
            state.subordinatesToShow = [];
            state.employee = payload;
            state.subordinates = payload.controls;
        },
        setSubordinates: (state, { payload }: PayloadAction<Employee[]>) => {
            state.subordinatesToShow = [
                ...state.subordinatesToShow,
                ...payload
            ];
            state.subordinates = payload.map((el) => el.controls).flat();
        }
    }
});

export const sliceActions = employeeSlice.actions;

export const selectEmployee = (state: RootState) => state.employee.employee;
export const selectSubordinates = (state: RootState) =>
    state.employee.subordinates;
export const selectSubordinatesToShow = (state: RootState) =>
    state.employee.subordinatesToShow;
export const selectSearch = (state: RootState) => state.employee.search;

export default employeeSlice.reducer;
