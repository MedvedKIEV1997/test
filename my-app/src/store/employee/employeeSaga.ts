import { all, call, put, select, takeLatest } from 'redux-saga/effects';
import axios, { AxiosResponse } from 'axios';

import * as actions from './employeeActions';
import { selectSubordinates, sliceActions } from './employeeSlice';
import { Employee } from './employeeTypes';

function* searchEmployees(action: ReturnType<typeof actions.searchEmployees>) {
    const {
        payload: { name }
    } = action;
    try {
        const { data }: AxiosResponse<Employee[]> = yield call(
            axios.get,
            `http://localhost:8000/employees/search/${name}`
        );
        yield put(sliceActions.setSearch(data));
    } catch (error) {
        console.log(error);
    }
}

function* getEmployee(action: ReturnType<typeof actions.getEmployee>) {
    const {
        payload: { id }
    } = action;

    try {
        const { data }: AxiosResponse<Employee> = yield call(
            axios.get,
            `http://localhost:8000/employees/${id}`
        );

        yield put(sliceActions.setEmployee(data));

        yield call(getSubordinates);
    } catch (error) {
        console.log(error);
    }
}

//I've tried to get the type
function* getSubordinates(): any {
    try {
        const subordinates: ReturnType<typeof selectSubordinates> =
            yield select(selectSubordinates);

        if (subordinates.length !== 0) {
            const response: AxiosResponse[] = yield all(
                subordinates.map((id) => {
                    const response = call(
                        axios.get,
                        `http://localhost:8000/employees/${id}`
                    );
                    return response;
                })
            );

            const subordinatesData: Employee[] = response.map((el) => el.data);

            yield put(sliceActions.setSubordinates(subordinatesData));

            yield* getSubordinates();
        }
    } catch (error) {
        console.log(error);
    }
}

function* watchSearch() {
    yield takeLatest(actions.searchEmployees.type, searchEmployees);
}

function* watchGetEmployee() {
    yield takeLatest(actions.getEmployee.type, getEmployee);
}

function* employeeSaga() {
    yield all([call(watchGetEmployee), call(watchSearch)]);
}

export default employeeSaga;
