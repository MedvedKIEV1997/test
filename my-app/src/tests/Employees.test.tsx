import { MemoryRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';

import Employees from '../components/Employees';
import { renderWithProviders } from './test-utils';
import { initialState } from '../redux/ducks/employee.duck';

describe('employees component tests', () => {
    const setup = () => {
        return {
            user: userEvent.setup(),
            ...renderWithProviders(
                <MemoryRouter>
                    <Employees />
                </MemoryRouter>,
                {
                    preloadedState: {
                        employee: {
                            ...initialState,
                            loading: false,
                            employee: {
                                name: 'Bob',
                                controls: [5, 6],
                                id: 11
                            },
                            subordinatesToShow: [
                                {
                                    name: 'Leam',
                                    controls: [],
                                    id: 5
                                },
                                {
                                    name: 'Caren',
                                    controls: [],
                                    id: 6
                                }
                            ]
                        }
                    }
                }
            )
        };
    };

    it('should match a snapshot', () => {
        const { asFragment } = setup();
        expect(asFragment()).toMatchSnapshot();
    });
});
