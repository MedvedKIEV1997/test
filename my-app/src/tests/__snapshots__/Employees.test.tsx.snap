// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`employees component tests should match a snapshot 1`] = `
<DocumentFragment>
  <div
    class="my-5 position-relative container"
  >
    <h5
      class="sc-bczRLJ jLIzaW"
    >
      Employee:
    </h5>
    <p
      class="sc-gsnTZi bvivqN"
    >
      Bob 
    </p>
    <h5
      class="sc-bczRLJ jLIzaW"
    >
      Subordinates:
    </h5>
    <p
      class="sc-gsnTZi bvivqN"
    >
      Leam, Caren
    </p>
    <button
      class="position-absolute top-0 end-0 btn btn-primary"
      type="button"
    >
      Copy this URL
    </button>
    <div
      class="toast-container position-absolute bottom-0 start-0 p-3"
    />
  </div>
</DocumentFragment>
`;
